File: scheme/ocaml/eval.ml
open Ast

exception RuntimeError of string

let rec set_args params args env = match params, args with
	| (Var name)::params', value::args' ->
		Environment.set env name value;
		set_args params' args' env
	| [Dotted (Var name)], lst -> Environment.set env name (List lst)
	| [], [] -> ()
	| (Var name)::_, [] -> raise (RuntimeError ("Argument " ^ name ^ " not given"))
	| _, _ -> raise (RuntimeError "Unable to set arguments")

let rec eval (e : expr) (env : environment) : expr =
	match e with
	| String _ | Integer _ | Bool _ | Closure _ | LibraryFunction _ | LibraryMacro _ -> e
	| Dotted _ -> raise (RuntimeError "Can't evaluate dotted expression")
	| Quoted e' -> e'
	| Var v -> (match Environment.get env v with
		| None -> raise (RuntimeError ("Unbound variable: " ^ v))
		| Some value -> value)
	| StatementList lst ->
		let rec eval_lst l env = match l with
		| [] -> raise (RuntimeError ("StatementList is empty"))
		| [hd] -> eval hd env
		| hd::tl -> let _ = eval hd env in eval_lst tl env
		in
		eval_lst lst env
	| List [] -> raise (RuntimeError "Can't evaluate empty list")
	| List(hd::tl) ->
		match eval hd env with
		| LibraryFunction lf -> lf (List.map (fun e -> eval e env) tl)
		| LibraryMacro lm -> lm tl env
		| Closure(params, env', body) ->
			let args = List.map (fun e -> eval e env) tl in
			let env'' = Environment.new_with_parent env' in
			set_args params args env'';
			eval body env''
		| _ -> raise (RuntimeError "This expression is not a function; it cannot be applied")

File: scheme/haskell/Eval.hs
module Eval(eval, setArgs) where

import Ast

eval :: Expr -> Environment -> IO Expr
eval (Integer i) _ = return $ Integer i
eval (String s) _ = return $ String s
eval (Bool b) _ = return $ Bool b
eval fn@(Closure _ _ _) _ = return fn
eval fn@(LibraryFunction _) _ = return fn
eval fn@(LibraryMacro _) _ = return fn

eval (Var x) env = do
    value <- getEnv env x
    case value of
        Nothing -> fail $ "Unbound variable: " ++ x
        Just val -> return val

eval (Quoted e) _ = return e

eval (Dotted _) _ = fail "Cannot evaluate dotted expression"

eval (StatementList []) _ = fail "Empty program"
eval (StatementList [hd]) env = eval hd env
eval (StatementList (hd:tl)) env = do
    eval hd env
    eval (StatementList tl) env

eval (List []) _ = fail "Cannot evaluate empty list"
eval (List (hd:tl)) env = do
    fn <- eval hd env
    case fn of
        LibraryFunction lf -> do
            args <- mapM (flip eval env) tl
            lf args
        LibraryMacro lm -> lm tl env
        Closure params savedEnv body -> do
            args <- mapM (flip eval env) tl
            newEnv <- makeEnv (Just savedEnv) Nothing
            setArgs params args newEnv
            eval body newEnv

setArgs :: [Expr] -> [Expr] -> Environment -> IO ()
setArgs (Var name:params) (value:args) env = do
    setEnv env name value
    setArgs params args env

setArgs [Dotted (Var name)] args env = setEnv env name (List args)
setArgs [] [] _ = return ()

setArgs (Var name:_) [] _ = fail $ "Argument " ++ name ++ " not given"
setArgs _ _ _ = fail "Unable to set arguments"

